File: java/src/main/java/sa/amnn/kit/resources/Deals.java
package sa.amnn.kit.resources;

import com.fasterxml.jackson.core.type.TypeReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import sa.amnn.kit.AmenClient;
import sa.amnn.kit.AmenClient.Part;
import sa.amnn.kit.AmenLifecycleError;
import sa.amnn.kit.Models.*;

public final class Deals {
    private final AmenClient c; private final DealActions actions;
    public Deals(AmenClient c) { this.c = c; this.actions = new DealActions(c, this); }
    public DealActions actions() { return actions; }

    public Deal create(CreateDeal body) { return c.post("/deals/", body, new TypeReference<>() {}); }
    public Deal get(String n) { return c.get("/deals/" + n, Map.of(), new TypeReference<>() {}); }
    public Deal update(String n, CreateDeal body) { return c.put("/deals/" + n, body, new TypeReference<>() {}); }
    public void delete(String n) { c.delete("/deals/" + n); }
    public Page<Deal> list(Map<String, String> params) { return c.page("/deals/", params, "deals", Deal.class); }
    public Deal setParties(String n, List<String> buyers, List<String> sellers) { return c.post("/deals/" + n + "/parties/", Map.of("buyers", buyers, "sellers", sellers), new TypeReference<>() {}); }
    public Deal setDeliveryAddress(String n, Address a) { return c.post("/deals/" + n + "/delivery-address", a, new TypeReference<>() {}); }
    public Deal setBillingAddress(String n, Address a) { return c.post("/deals/" + n + "/billing-address", a, new TypeReference<>() {}); }
    public List<String> allowedPaymentMethods(String n) {
        Map<String, List<String>> r = c.get("/deals/" + n + "/allowed-payment-methods/", Map.of(), new TypeReference<>() {});
        return r == null ? List.of() : r.getOrDefault("payment_methods", List.of());
    }

    /** POST /deals/{n}/action/* — every method returns the updated Deal (or a Checkout for online payment). */
    public static final class DealActions {
        /** Which statuses each action may be called from (docs/02-deal-lifecycle.md). */
        public static final Map<String, Set<String>> ALLOWED_FROM = Map.ofEntries(
            Map.entry("submit", Set.of("draft")), Map.entry("approve", Set.of("requested")),
            Map.entry("make-payment-wallet", Set.of("payment_pending")), Map.entry("make-payment-online", Set.of("payment_pending")),
            Map.entry("execution-start", Set.of("paid")), Map.entry("execution-complete", Set.of("executing")), Map.entry("complete", Set.of("executed")),
            Map.entry("transfer-seller-amount", Set.of("completed")), Map.entry("dispute", Set.of("completed")),
            Map.entry("dispute-approve", Set.of("disputed")), Map.entry("dispute-decline", Set.of("disputed")),
            Map.entry("cancel", Set.of("draft", "requested", "payment_pending", "paid", "executing")));
        private final AmenClient c; private final Deals deals;
        DealActions(AmenClient c, Deals deals) { this.c = c; this.deals = deals; }

        private <T> T act(String n, String action, Object json, List<Part> form, boolean check, TypeReference<T> type) {
            if (check) {
                String status = deals.get(n).status();
                if (!ALLOWED_FROM.get(action).contains(status)) throw new AmenLifecycleError("action '" + action + "' is not allowed from status '" + status + "' (allowed: " + ALLOWED_FROM.get(action) + ")");
            }
            return form != null ? c.postForm("/deals/" + n + "/action/" + action, form, type) : c.post("/deals/" + n + "/action/" + action, json == null ? Map.of() : json, type);
        }
        private static final TypeReference<Deal> DEAL = new TypeReference<>() {};
        public Deal submit(String n) { return act(n, "submit", null, null, true, DEAL); }
        public Deal approve(String n, String price) { return act(n, "approve", price == null ? Map.of() : Map.of("price", price), null, true, DEAL); }
        public Deal payWithWallet(String n) { return act(n, "make-payment-wallet", null, null, true, DEAL); }
        public Checkout payOnline(String n, String paymentMethod) { return act(n, "make-payment-online", Map.of("payment_method", paymentMethod), null, true, new TypeReference<>() {}); }
        public Deal executionStart(String n) { return act(n, "execution-start", null, null, true, DEAL); }
        public Deal executionComplete(String n) { return act(n, "execution-complete", null, null, true, DEAL); }
        public Deal complete(String n) { return act(n, "complete", null, null, true, DEAL); }
        public Deal transferSellerAmount(String n) { return act(n, "transfer-seller-amount", null, null, true, DEAL); }
        public Deal cancel(String n, String dealParty, int reason, String comment) { return act(n, "cancel", Map.of("deal_party", dealParty, "reason", reason, "comment", comment), null, true, DEAL); }
        public Deal dispute(String n, int reason, String comment, List<Part> attachments) {
            List<Part> form = new ArrayList<>(List.of(Part.text("reason", String.valueOf(reason)), Part.text("comment", comment)));
            for (int i = 0; i < attachments.size(); i++) { Part a = attachments.get(i); form.add(Part.file("attachment_" + (i + 1), a.filename(), a.file(), a.mimeType())); }
            return act(n, "dispute", null, form, true, DEAL);
        }
        public Deal disputeApprove(String n, int reason, String comment) { return act(n, "dispute-approve", null, List.of(Part.text("reason", String.valueOf(reason)), Part.text("comment", comment)), true, DEAL); }
        public Deal disputeDecline(String n, int reason, String comment) { return act(n, "dispute-decline", null, List.of(Part.text("reason", String.valueOf(reason)), Part.text("comment", comment)), true, DEAL); }
    }
}

File: kotlin/amen-client/src/main/kotlin/sa/amnn/kit/resources/Deals.kt
package sa.amnn.kit.resources

import kotlinx.serialization.json.*
import sa.amnn.kit.*

class Deals(private val c: AmenClient) {
    val actions = DealActions(c, this)
    suspend fun create(body: CreateDeal) = c.request("POST", "/deals/", Deal.serializer(), body = json.encodeToJsonElement(body))!!
    suspend fun get(n: String) = c.request("GET", "/deals/$n", Deal.serializer())!!
    suspend fun update(n: String, body: CreateDeal) = c.request("PUT", "/deals/$n", Deal.serializer(), body = json.encodeToJsonElement(body))!!
    suspend fun delete(n: String) { c.request("DELETE", "/deals/$n", JsonElement.serializer()) }
    suspend fun list(params: Map<String, String?> = emptyMap()) = c.page("/deals/", params, "deals", Deal.serializer())
    suspend fun setParties(n: String, buyers: List<String>, sellers: List<String>) =
        c.request("POST", "/deals/$n/parties/", Deal.serializer(), body = buildJsonObject { put("buyers", JsonArray(buyers.map(::JsonPrimitive))); put("sellers", JsonArray(sellers.map(::JsonPrimitive))) })!!
    suspend fun setDeliveryAddress(n: String, a: Address) = c.request("POST", "/deals/$n/delivery-address", Deal.serializer(), body = json.encodeToJsonElement(a))!!
    suspend fun setBillingAddress(n: String, a: Address) = c.request("POST", "/deals/$n/billing-address", Deal.serializer(), body = json.encodeToJsonElement(a))!!
    suspend fun allowedPaymentMethods(n: String): List<String> =
        c.request("GET", "/deals/$n/allowed-payment-methods/", JsonElement.serializer())?.jsonObject?.get("payment_methods")?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList()
}

/** POST /deals/{n}/action/* — every method returns the updated Deal (or a Checkout for online payment). */
class DealActions(private val c: AmenClient, private val deals: Deals) {
    companion object {
        /** Which statuses each action may be called from (docs/02-deal-lifecycle.md). */
        val ALLOWED_FROM = mapOf(
            "submit" to setOf("draft"), "approve" to setOf("requested"),
            "make-payment-wallet" to setOf("payment_pending"), "make-payment-online" to setOf("payment_pending"),
            "execution-start" to setOf("paid"), "execution-complete" to setOf("executing"), "complete" to setOf("executed"),
            "transfer-seller-amount" to setOf("completed"), "dispute" to setOf("completed"),
            "dispute-approve" to setOf("disputed"), "dispute-decline" to setOf("disputed"),
            "cancel" to setOf("draft", "requested", "payment_pending", "paid", "executing"),
        )
    }
    private suspend fun <T> act(n: String, action: String, serializer: kotlinx.serialization.KSerializer<T>, body: JsonElement? = null, form: List<AmenClient.Part>? = null, check: Boolean = true): T {
        if (check) {
            val status = deals.get(n).status
            if (status !in ALLOWED_FROM.getValue(action)) throw AmenLifecycleError("action '$action' is not allowed from status '$status' (allowed: ${ALLOWED_FROM.getValue(action).joinToString(", ")})")
        }
        return c.request("POST", "/deals/$n/action/$action", serializer, body = if (form == null) (body ?: JsonObject(emptyMap())) else null, form = form)!!
    }
    private fun form(reason: Int, comment: String) = listOf(AmenClient.Part("reason", reason.toString()), AmenClient.Part("comment", comment))

    suspend fun submit(n: String, check: Boolean = true) = act(n, "submit", Deal.serializer(), check = check)
    suspend fun approve(n: String, price: String? = null, check: Boolean = true) = act(n, "approve", Deal.serializer(), buildJsonObject { price?.let { put("price", it) } }, check = check)
    suspend fun payWithWallet(n: String, check: Boolean = true) = act(n, "make-payment-wallet", Deal.serializer(), check = check)
    suspend fun payOnline(n: String, paymentMethod: String = "mada", check: Boolean = true) = act(n, "make-payment-online", Checkout.serializer(), buildJsonObject { put("payment_method", paymentMethod) }, check = check)
    suspend fun executionStart(n: String, check: Boolean = true) = act(n, "execution-start", Deal.serializer(), check = check)
    suspend fun executionComplete(n: String, check: Boolean = true) = act(n, "execution-complete", Deal.serializer(), check = check)
    suspend fun complete(n: String, check: Boolean = true) = act(n, "complete", Deal.serializer(), check = check)
    suspend fun transferSellerAmount(n: String, check: Boolean = true) = act(n, "transfer-seller-amount", Deal.serializer(), check = check)
    suspend fun cancel(n: String, dealParty: String, reason: Int, comment: String, check: Boolean = true) =
        act(n, "cancel", Deal.serializer(), buildJsonObject { put("deal_party", dealParty); put("reason", reason); put("comment", comment) }, check = check)
    suspend fun dispute(n: String, reason: Int, comment: String, attachments: List<AmenClient.Part> = emptyList(), check: Boolean = true) =
        act(n, "dispute", Deal.serializer(), form = form(reason, comment) + attachments.mapIndexed { i, a -> AmenClient.Part("attachment_${i + 1}", file = a.file, filename = a.filename, mimeType = a.mimeType) }, check = check)
    suspend fun disputeApprove(n: String, reason: Int, comment: String, check: Boolean = true) = act(n, "dispute-approve", Deal.serializer(), form = form(reason, comment), check = check)
    suspend fun disputeDecline(n: String, reason: Int, comment: String, check: Boolean = true) = act(n, "dispute-decline", Deal.serializer(), form = form(reason, comment), check = check)
}
